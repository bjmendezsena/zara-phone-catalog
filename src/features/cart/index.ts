export * from "./context";
export * from "./helpers";
export * from "./types";
export * from "./components";
