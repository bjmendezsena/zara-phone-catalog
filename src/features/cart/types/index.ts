export type CartItem = {
  productId: string;
  color: string;
  storage: string;
  brand: string;
  name: string;
  price: number;
  imageUrl: string;
  quantity?: number;
};
