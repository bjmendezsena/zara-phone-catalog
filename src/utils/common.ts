export const isClientSide = (): boolean => {
  return typeof window !== "undefined";
};

export const getValue = <T extends Record<string, unknown>>(
  source: T,
  key: string
): unknown => {
  const arrKeys = key.split(".") as string[];
  const firstKey = arrKeys.shift() || "";
  const newSource = source[firstKey as keyof typeof source];

  if (arrKeys.length > 0 && newSource && typeof newSource === "object") {
    return getValue(newSource as Record<string, unknown>, arrKeys.join("."));
  }

  return newSource;
};
