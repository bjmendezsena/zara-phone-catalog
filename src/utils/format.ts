export const formatCurrency = (amount: number): string => {
  return `${amount.toFixed(2)} EUR`;
};

export function formatString(fmt: string, ...args: any[]): string {
  const re = /(%?)(%([ojds]))/g;

  if (args.length) {
    fmt = fmt.replace(re, (match, escaped, _ptn, flag) => {
      let arg = args.shift();

      switch (flag) {
        case "o":
          if (Array.isArray(arg)) {
            arg = JSON.stringify(arg);
          }
          break;
        case "s":
          arg = "" + arg;
          break;
        case "d":
          arg = Number(arg);
          break;
        case "j":
          arg = JSON.stringify(arg);
          break;
        default:
          break;
      }

      if (!escaped) {
        return arg;
      }

      args.unshift(arg);
      return match;
    });
  }

  // arguments remain after formatting
  if (args.length) {
    fmt += " " + args.join(" ");
  }

  // update escaped %% values
  fmt = fmt.replace(/%{2,2}/g, "%");

  return "" + fmt;
}
