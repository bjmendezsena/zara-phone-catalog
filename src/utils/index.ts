export * from "./cn";
export * from "./uuid";
export * from "./create-context";
export * from "./json";
export * from "./common";
export * from "./format";
