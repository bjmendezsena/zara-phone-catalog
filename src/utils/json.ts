export const isValidJSON = (
  value: unknown
): value is Record<string, unknown> => {
  if (typeof value !== "string") return false;
  try {
    JSON.parse(value);
    return true;
  } catch {
    return false;
  }
};
