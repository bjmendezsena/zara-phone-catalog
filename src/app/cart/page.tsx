"use client";
import { useMemo } from "react";
import Link from "next/link";
import { formatCurrency } from "@/utils";
import { useCartContext, ProductCartItem } from "@/features/cart";

const CartRoute = () => {
  const { items, removeItem } = useCartContext();

  const totalItems = useMemo(() => {
    return items.reduce((total, item) => total + (item.quantity || 1), 0);
  }, [items]);

  const totalPrice = useMemo(() => {
    return items.reduce(
      (total, item) => total + item.price * (item.quantity || 1),
      0
    );
  }, [items]);

  if (items.length === 0) {
    return (
      <div className='flex flex-col items-center justify-center min-h-screen p-4'>
        <h1 className='text-2xl font-light text-primary mb-4 uppercase tracking-wide'>
          Your cart is empty
        </h1>
        <p className='text-secondary'>Add some products to get started</p>
      </div>
    );
  }

  return (
    <div className='flex flex-col h-screen'>
      {/* Header */}
      <div className='flex-shrink-0 p-4 md:p-6 '>
        <h1 className='text-2xl font-light text-primary uppercase tracking-wide'>
          Cart ({totalItems})
        </h1>
      </div>

      {/* Scrollable cart items */}
      <div className='flex-1 overflow-y-auto pb-32'>
        <div className='space-y-4'>
          {items.map((item, index) => (
            <ProductCartItem
              key={`${item.productId}-${item.color}-${item.storage}-${index}`}
              item={item}
              onRemove={() => removeItem(item.productId)}
            />
          ))}
        </div>
      </div>

      {/* Sticky footer with buttons */}
      <div className='fixed bottom-0 left-0 right-0 bg-white p-2 pb-6 z-50'>
        <div className=''>
          {/* Total */}
          <div className='flex justify-between items-center mb-2 md:hidden'>
            <span className='text-lg uppercase tracking-wide'>TOTAL</span>
            <span className='text-lg '>{formatCurrency(totalPrice)}</span>
          </div>

          <div className='flex flex-row gap-2 sm:gap-4'>
            <Link
              href={"/"}
              className='text-center flex-1 py-3 cursor-pointer border border-gray-300 text-sm uppercase tracking-wide text-secondary hover:border-primary hover:text-primary transition-colors'
            >
              CONTINUE SHOPPING
            </Link>
            <div className='flex gap-6 justify-center items-center mb-2 flex-1'>
              <span className='text-lg font-medium uppercase tracking-wide'>
                TOTAL
              </span>
              <span className='text-lg font-medium'>
                {formatCurrency(totalPrice)}
              </span>
            </div>
            <button className='flex-1 py-3  bg-primary text-white text-sm uppercase tracking-wide cursor-pointer'>
              PAY
            </button>
          </div>
        </div>
      </div>
    </div>
  );
};

export default CartRoute;
