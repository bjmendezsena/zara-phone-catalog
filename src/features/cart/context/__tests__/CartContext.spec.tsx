import { render, waitFor, getHtml, userEvent } from "@/tests/utils";
import { CartProvider, CartItem, useCartContext } from "@/features/cart";

type TestComponentProps = {
  items: CartItem[];
};

const mockItem: CartItem = {
  productId: "1",
  name: "Test Product",
  price: 100,
  color: "Red",
  storage: "64GB",
  quantity: 1,
  brand: "Test Brand",
  imageUrl: "http://example.com/image.jpg",
};

describe(`<${CartProvider.name}/>`, () => {
  const TestComponent = ({ items }: TestComponentProps) => {
    const {
      items: cartItems,
      addItem,
      removeItem,
      clearCart,
      totalItems,
    } = useCartContext();
    return (
      <div>
        <h1>Cart Items: {cartItems.length}</h1>
        <button onClick={() => addItem(items[0])}>Add Item</button>
        <button onClick={() => removeItem(items[0].productId)}>
          Remove Item
        </button>
        <button onClick={clearCart}>Clear Cart</button>
        <p>Total Items: {totalItems}</p>
      </div>
    );
  };

  const getComponent = (props: TestComponentProps) =>
    render(
      <CartProvider>
        <TestComponent {...props} />
      </CartProvider>
    );
  it("should render with initial state", () => {
    const items: CartItem[] = [];
    const { getByText } = getComponent({ items });

    expect(getByText("Cart Items: 0")).toBeInTheDocument();
    expect(getByText("Total Items: 0")).toBeInTheDocument();
  });
  it("should add an item to the cart", async () => {
    const { getByText } = getComponent({ items: [mockItem] });

    const addButton = getByText("Add Item");
    await userEvent.click(addButton);

    expect(getByText("Cart Items: 1")).toBeInTheDocument();
    expect(getByText("Total Items: 1")).toBeInTheDocument();
  });

  it("should remove an item from the cart", async () => {
    const { getByText } = getComponent({ items: [mockItem] });

    const addButton = getByText("Add Item");
    await userEvent.click(addButton);

    const removeButton = getByText("Remove Item");
    await userEvent.click(removeButton);

    expect(getByText("Cart Items: 0")).toBeInTheDocument();
    expect(getByText("Total Items: 0")).toBeInTheDocument();
  });
  it("should clear the cart", async () => {
    const { getByText } = getComponent({ items: [mockItem] });

    const addButton = getByText("Add Item");
    await userEvent.click(addButton);

    const clearButton = getByText("Clear Cart");
    await userEvent.click(clearButton);

    expect(getByText("Cart Items: 0")).toBeInTheDocument();
    expect(getByText("Total Items: 0")).toBeInTheDocument();
  });
  it("should handle multiple items correctly", async () => {
    const items: CartItem[] = [{ ...mockItem, productId: "1" }];
    const { getByText } = getComponent({ items });

    expect(getByText("Cart Items: 0")).toBeInTheDocument();
    expect(getByText("Total Items: 0")).toBeInTheDocument();

    await userEvent.click(getByText("Add Item"));

    await waitFor(() => {
      expect(getByText("Cart Items: 1")).toBeInTheDocument();
      expect(getByText("Total Items: 1")).toBeInTheDocument();
    });
  });

  it("should render the cart items correctly", async () => {
    const items: CartItem[] = [mockItem];
    const { getByText, container } = getComponent({ items });

    await userEvent.click(getByText("Add Item"));

    expect(getByText("Cart Items: 1")).toBeInTheDocument();
    expect(getByText("Total Items: 1")).toBeInTheDocument();
    expect(getHtml(container)).toMatchSnapshot();
  });
});
