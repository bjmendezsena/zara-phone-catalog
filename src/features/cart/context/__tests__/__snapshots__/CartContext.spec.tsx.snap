// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<CartProvider/> should render the cart items correctly 1`] = `
"<div>
  <div>
    <h1>
      Cart Items: 
      1
    </h1>
    <button>
      Add Item
    </button>
    <button>
      Remove Item
    </button>
    <button>
      Clear Cart
    </button>
    <p>
      Total Items: 
      1
    </p>
  </div>
</div>"
`;
